### PR TITLE
Active set size calculation code reuse

### PR DESCRIPTION
**Pull request by gavraz**

### CODE REVIEW

File: activation/atxdb.go [hunk 1]
```diff
@@ -68,6 +73,93 @@ func (db *ActivationDb) ProcessAtx(atx *types.ActivationTx) {
...
+
+	epoch := layer.GetEpoch(layersPerEpoch)
+	if epoch == 0 {
+		return nil, errors.New("tried to retrieve active set for epoch 0")
+	}
+
+	firstLayerOfPrevEpoch := types.LayerID(epoch-1) * types.LayerID(layersPerEpoch)
+
+	// build a map of all blocks on the current layer
+	mp, err := db.blocksProvider.GetGoodPatternBlocks(layer)
```
Comment (zalmen) [inline activation/atxdb.go]: you don't really need the blockProvider, just pass the starting blocks to this method. It overcomplicates the ADB for nothing

File: activation/atxdb.go [hunk 2]
```diff
@@ -68,6 +73,93 @@ func (db *ActivationDb) ProcessAtx(atx *types.ActivationTx) {
...
+		return nil
+	}
+
+	return traversalFunc
+}
+
+// ActiveSetForLayerConsensusView - returns the active set size that matches the view of the contextually valid blocks in the provided layer
+func (db *ActivationDb) ActiveSetForLayerConsensusView(layer types.LayerID, layersPerEpoch uint16) (map[string]struct{}, error) {
+
+	epoch := layer.GetEpoch(layersPerEpoch)
```
Comment (zalmen) [inline activation/atxdb.go]: This is the only place where you use `layer` and `layersPerEpoch`. Why don't you just pass the method the epoch instead?

File: activation/atxdb.go [hunk 3]
```diff
@@ -68,6 +73,93 @@ func (db *ActivationDb) ProcessAtx(atx *types.ActivationTx) {
...
+		}
+
+		return nil
+	}
+
+	return traversalFunc
+}
+
+// ActiveSetForLayerConsensusView - returns the active set size that matches the view of the contextually valid blocks in the provided layer
+func (db *ActivationDb) ActiveSetForLayerConsensusView(layer types.LayerID, layersPerEpoch uint16) (map[string]struct{}, error) {
```
Comment (zalmen) [inline activation/atxdb.go]: this should be a generic function, rename to something without "ConsensusView"

File: activation/atxdb.go [hunk 4]
```diff
@@ -82,46 +174,16 @@ func (db *ActivationDb) CalcActiveSetFromView(a *types.ActivationTx) (uint32, er
...
-					atx.ShortId(), atx.TargetEpoch(db.LayersPerEpoch), pubEpoch)
-				continue
-			}
-			counter++
-			db.log.Debug("atx %v (epoch %d) found traversing in block %x (epoch %d)",
-				atx.ShortId(), atx.TargetEpoch(db.LayersPerEpoch), blk.Id, blk.LayerIndex.GetEpoch(db.LayersPerEpoch))
-		}
-		return nil
-	}
+	countedAtxs := make(map[string]types.AtxId)
```
Comment (zalmen) [inline activation/atxdb.go]: you can re-use `ActiveSetForLayerConsensusView`, once you make it general purposed

**Review by zalmen — CHANGES_REQUESTED** [review]

File: activation/atxdb.go [hunk 5]
```diff
@@ -82,46 +174,16 @@ func (db *ActivationDb) CalcActiveSetFromView(a *types.ActivationTx) (uint32, er
...
 	firstLayerOfLastEpoch := types.LayerID(countingEpoch) * types.LayerID(db.LayersPerEpoch)
 
-	traversalFunc := func(blkh *types.BlockHeader) error {
-		blk, err := db.meshDb.GetBlock(blkh.Id)
-		if err != nil {
-			db.log.With().Error("cannot validate atx, block wasn't found", log.AtxId(a.ShortId()), log.BlockId(uint64(blkh.Id)))
-			return err
-		}
-		//skip blocks not from atx epoch
-		if blk.LayerIndex.GetEpoch(db.LayersPerEpoch) != countingEpoch {
```
Comment (noamnelke) [inline activation/atxdb.go:93]: you've omitted this check in your code, which has a performance cost, but also means you're counting ATXs that may not have been referenced in the epoch you're supposed to be counting them in.

File: activation/atxdb.go (hunk 1 above)
Comment (gavraz) [reply activation/atxdb.go]: Fixed

File: activation/atxdb.go (hunk 2 above)
Comment (gavraz) [reply activation/atxdb.go]: fixed

File: activation/atxdb.go (hunk 3 above)
Comment (gavraz) [reply activation/atxdb.go]: Fixed

File: activation/atxdb.go (hunk 4 above)
Comment (gavraz) [reply activation/atxdb.go]: Fixed

File: activation/atxdb.go (hunk 5 above)
Comment (gavraz) [reply activation/atxdb.go:93]: Fixed

**Review by zalmen — APPROVED** [review]

File: activation/atxdb.go [hunk 6]
```diff
@@ -68,6 +68,87 @@ func (db *ActivationDb) ProcessAtx(atx *types.ActivationTx) {
 	}
 }
 
+func (db *ActivationDb) createTraversalActiveSetCounterFunc(countedAtxs map[string]types.AtxId, penalties map[string]struct{}, layersPerEpoch uint16, epoch types.EpochId) func(b *types.Block) error {
```
Comment (antonlerner) [inline activation/atxdb.go:68]: why do we need to create a function? name the function and use it where you want to

File: mesh/mesh_test.go [hunk 7]
```diff
@@ -276,91 +274,3 @@ func TestLayers_OrphanBlocks(t *testing.T) {
...
-	return
-}
-
-func rndStr() string {
-	a := make([]byte, 4)
-	rand.Read(a)
-	return string(a)
-}
-
-func TestMesh_ActiveSetForLayerView(t *testing.T) {
```
Comment (antonlerner) [inline mesh/mesh_test.go:316]: we need to tests this code, make sure you have these tests somewhere else

Comment (gavraz) [reply mesh/mesh_test.go:316]: activation_test

File: activation/atxdb.go [hunk 8]
```diff
@@ -65,6 +65,87 @@ func (db *ActivationDb) ProcessAtx(atx *types.ActivationTx) {
...
+}
+
+// CalcActiveSetSize - returns the active set size that matches the view of the contextually valid blocks in the provided layer
+func (db *ActivationDb) CalcActiveSetSize(epoch types.EpochId, blocks map[types.BlockID]struct{}) (map[string]struct{}, error) {
+
+	if epoch == 0 {
+		return nil, errors.New("tried to retrieve active set for epoch 0")
+	}
+
+	firstLayerOfPrevEpoch := types.LayerID(epoch-1) * types.LayerID(db.LayersPerEpoch)
```
Comment (noamnelke) [inline activation/atxdb.go:129]: could have done `(epoch - 1).FirstLayer(db.LayersPerEpoch)` which returns a `LayerID`

**Pull request merged**